src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -425,52 +511,127 @@ public class SqlMutationEngine : IMutationEngine
                 }
 
-                if (context.OperationType is EntityActionOperation.Insert)
-                {
-                    if (mutationResultRow is null)
-                    {
-                        // Ideally this case should not happen, however may occur due to unexpected reasons,
-                        // like the DbDataReader being null. We throw an exception
-                        // which will be returned as an Unexpected Internal Server Error
-                        throw new DataApiBuilderException(
-                            message: "An unexpected error occurred while trying to execute the query.",
-                            statusCode: HttpStatusCode.InternalServerError,
-                            subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedError);
-                    }
+                string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow!.Columns);
 
-                    if (mutationResultRow.Columns.Count == 0)
+                // When read permission is configured without a database policy, a subsequent select query will not be executed.
+                // So, if the read action has include/exclude fields configured, additional fields present in the response
+                // need to be removed.
+                if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                {
+                    HashSet<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read).ToHashSet();
+                    foreach (string columnInResponse in mutationResultRow!.Columns.Keys)
                     {
-                        throw new DataApiBuilderException(
-                            message: "Could not insert row with given values.",
-                            statusCode: HttpStatusCode.Forbidden,
-                            subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure
-                            );
+                        if (!allowedExposedColumns.Contains(columnInResponse))
+                        {
+                            mutationResultRow!.Columns.Remove(columnInResponse);
+                        }
                     }
+                }
 
-                    string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow.Columns);
-                    // location will be updated in rest controller where httpcontext is available
-                    return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(mutationResultRow.Columns).Value);
+                if (context.OperationType is EntityActionOperation.Insert)
+                {
+                    return ConstructCreatedResultResponse(mutationResultRow!.Columns, selectOperationResponse, primaryKeyRoute, isReadPermissionConfiguredForRole, isDatabasePolicyDefinedForReadAction);
```

Comment:
Would be helpful to mention in a comment why you are adding primary key for insert operation and not for the others.